addon/utils/serializer-partial-model-handler.js

```diff
@@ -0,0 +1,33 @@
+export default {
+  preNormalize: function(modelClass, resourceHash, context, normalizePartialRelationshipStrategy) {
+    if (modelClass._isPartialModel) {
+      partialDescriptors(modelClass).forEach(descriptor => {
+        normalizePartialRelationshipStrategy(modelClass, resourceHash, descriptor, context);
+      });
+    }
+  },
+
+  postSerialize: function(snapshot, serializedHash, copyAttributesFromPartialToParentStrategy,
+                          modifyHashBasedOnDescriptorKeyStrategy) {
+    partialDescriptors(snapshot).forEach(descriptor => {
+      let partial = snapshot.belongsTo(descriptor.key);
+
+      modifyHashBasedOnDescriptorKeyStrategy(serializedHash, descriptor.key);
+
+      if (partial) {
+        let partialHash = partial.record.serialize();
+        copyAttributesFromPartialToParentStrategy(serializedHash, partialHash);
+      }
+    });
+  }
+};
+
+function partialDescriptors(modelClass) {
+  if (modelClass._partialDescriptors) {
```

Comment:
We need to extract partial descriptors either from modelClass or snapshot, so that's helper function for some unified way of accessing it and returning empty array if there are no partial descriptors so that it doesn't break in any case. Maybe the argument name could be `modelClassOrSnapshot` so it's closer to what it actually might be.